src/election_works/voting_methods.clj

```
@@ -0,0 +1,81 @@
+(ns election-works.voting-methods
+  (:require [clojure.tools.logging :as log]
+            [election-works.utils :refer [maybe-assoc]]))
+
+(defn assoc-props
+  [{:keys [ballot-request-deadline
+           early-voting-start early-voting-end]}
+    voting-method]
+  (case (:type voting-method)
+    :by-mail
+    (maybe-assoc voting-method
+                 :ballot-request-deadline-received ballot-request-deadline)
+    :early-voting
+    (-> voting-method
+        (maybe-assoc :start early-voting-start)
+        (maybe-assoc :end early-voting-end))
+    voting-method))
+
+(defn merge-voting-methods
+  [defaults election-props]
+  (map (partial assoc-props election-props) defaults))
```

To me it feels like this part of the code relies on details of the import spreadsheet that I don't think we want to creep into the main codebase.

A couple of thoughts:

1. I think we want to throw an error if the import includes data for unsupported voting-methods (e.g. if if has early-voting-start or end when there is no early-voting method).

    1a. actually this brings up a question about deadlines -- it seems like different jurisdictions might have different deadlines; (a) I might be wrong about this, and (b) I suppose that could be an exceptional case where we need to think of those as different elections. But the data model would support different deadlines by jurisdiction, so it would be nice if we could figure out a reasonable way to do that at some point in the future, perhaps when we tackle a richer election import interface.

2. I still think it might be more flexible (and possibly cleaner) to build partially filled voting methods from the election import, and then `merge` them into the authority voting methods (similar to the previous `merge-voting-methods` function). We could still `throw` if a voting method shows up in the import that doesn't exist in the authority.

I remember you talking about issues with accidentally removing or adding voting methods with that function. If we wanted to do this, I would suggest something like this:

```clj
(-> (filter-allowed-voting-methods base-voting-methods #{:by-mail :in-person :whatever})
    ;; throw if there are methods that are not in base-voting-methods
    (merge-voting-methods election-deadline-vms)
    (select-primary-voting-method :by-mail))
```

which is almost exactly what you have now, but with the filtering done before the merging, so you can tell if the election-specific vms include bad data and throw if they try to expand voting-methods.

---

I went down the road of doing this, but now I remember why I changed the whole merge in the first place.  Your state authority voting methods and election authority voting methods don't have to contain the same voting-method types.  So, maybe the state-authority supports early-voting, in-person, and by-mail, but the local authority is all-mail.   So you specify early-voting dates for the state authority, but the local authority only supports by-mail, and so when you do this merge, you are stuck.

---

Has product talked about this sort of situation? Initially it feels like relaxing the import spec is a bad idea . . . would it work to throw an exception if the election supplies data that _none_ of the authority voting methods use? I've definitely had imports where the spreadsheet included data for vms that the state authority doesn't handle, and that threw an error -- but maybe the research team is actually OK with us ignoring that bad data?

If we needed to restrict the election vms to only the ones the authority supports, we could do something like this:
```clj
(let [authority-hash (->hash authority-vms)
      election-hash (select-keys (->hash election-vms) (keys authority-hash))])
```

---

OK, I talked to Esha, and we agreed that we can throw an exception if _none_ of the authority voting methods used a piece of data.